File: rb/test/component_group_uptime_direct_test.rb
# ComponentGroupUptime direct test

require "minitest/autorun"
require "json"
require_relative "../Statuspage_sdk"
require_relative "runner"

class ComponentGroupUptimeDirectTest < Minitest::Test
  def test_direct_load_component_group_uptime
    setup = component_group_uptime_direct_setup({ "id" => "direct01" })
    _should_skip, _reason = Runner.is_control_skipped("direct", "direct-load-component_group_uptime", setup[:live] ? "live" : "unit")
    if _should_skip
      skip(_reason || "skipped via sdk-test-control.json")
      return
    end
    if setup[:live]
      skip "live direct-load needs real ID — set *_ENTID env var with real IDs to run"
      return
    end
    client = setup[:client]

    params = {}
    query = {}
    unless setup[:live]
      params["id"] = "direct01"
      params["page_id"] = "direct02"
    end

    result = client.direct({
      "path" => "pages/{page_id}/component-groups/{id}/uptime",
      "method" => "GET",
      "params" => params,
      "query" => query,
    })
    if setup[:live]
      # Live mode is lenient: synthetic IDs frequently 4xx. Skip rather
      # than fail when the load endpoint isn't reachable with the IDs
      # we can construct from setup.idmap.
      if !result["err"].nil?
        skip("load call failed (likely synthetic IDs against live API): #{result["err"]}")
        return
      end
      unless result["ok"]
        skip("load call not ok (likely synthetic IDs against live API)")
        return
      end
      status = Helpers.to_int(result["status"])
      if status < 200 || status >= 300
        skip("expected 2xx status, got #{status}")
        return
      end
    else
      assert_nil result["err"]
      assert result["ok"]
      assert_equal 200, Helpers.to_int(result["status"])
      assert !result["data"].nil?
      if result["data"].is_a?(Hash)
        assert_equal "direct01", result["data"]["id"]
      end
      assert_equal 1, setup[:calls].length
    end
  end

end


def component_group_uptime_direct_setup(mockres)
  Runner.load_env_local

  calls = []

  env = Runner.env_override({
    "STATUSPAGE_TEST_COMPONENT_GROUP_UPTIME_ENTID" => {},
    "STATUSPAGE_TEST_LIVE" => "FALSE",
    "STATUSPAGE_APIKEY" => "NONE",
  })

  live = env["STATUSPAGE_TEST_LIVE"] == "TRUE"

  if live
    merged_opts = {
      "apikey" => env["STATUSPAGE_APIKEY"],
    }
    client = StatuspageSDK.new(merged_opts)
    return {
      client: client,
      calls: calls,
      live: true,
      idmap: {},
    }
  end

  mock_fetch = ->(url, init) {
    calls.push({ "url" => url, "init" => init })
    return {
      "status" => 200,
      "statusText" => "OK",
      "headers" => {},
      "json" => ->() {
        if !mockres.nil?
          return mockres
        end
        return { "id" => "direct01" }
      },
      "body" => "mock",
    }, nil
  }

  client = StatuspageSDK.new({
    "base" => "http://localhost:8080",
    "system" => {
      "fetch" => mock_fetch,
    },
  })

  {
    client: client,
    calls: calls,
    live: false,
    idmap: {},
  }
end

File: js/test/entity/component_group_uptime/ComponentGroupUptimeDirect.test.js

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

const { test, describe } = require('node:test')
const assert = require('node:assert')


const { StatuspageSDK } = require('../../..')

const {
  envOverride,
} = require('../../utility')


describe('ComponentGroupUptimeDirect', async () => {

  test('direct-exists', async () => {
    const sdk = new StatuspageSDK({
      // Concrete base: a live construction must satisfy any server
      // variables a templated base URL declares; overriding base with a
      // literal (as the direct flow tests do) sidesteps the requirement.
      base: 'http://localhost:8080',
      system: { fetch: async () => ({}) }
    })
    assert('function' === typeof sdk.direct)
    assert('function' === typeof sdk.prepare)
  })


  test('direct-load-component_group_uptime', async () => {
    const setup = directSetup({ id: 'direct01' })
    const { client, calls } = setup

    const params = {}
    if (!setup.live) {
      params.id = 'direct01'
      params.page_id = 'direct02'
    }

    const result = await client.direct({
      path: 'pages/{page_id}/component-groups/{id}/uptime',
      method: 'GET',
      params,
    })

    assert(result.ok === true)
    assert(result.status === 200)
    assert(null != result.data)

    if (!setup.live) {
      assert(result.data.id === 'direct01')
      assert(calls.length === 1)
      assert(calls[0].init.method === 'GET')
      assert(calls[0].url.includes('direct01'))
      assert(calls[0].url.includes('direct02'))
    }
  })

})



function directSetup(mockres) {
  const calls = []

  const env = envOverride({
    'STATUSPAGE_TEST_COMPONENT_GROUP_UPTIME_ENTID': {},
    'STATUSPAGE_TEST_LIVE': 'FALSE',
    'STATUSPAGE_APIKEY': 'NONE',
  })

  const live = 'TRUE' === env.STATUSPAGE_TEST_LIVE

  if (live) {
    const client = new StatuspageSDK({
      apikey: env.STATUSPAGE_APIKEY,
    })

    let idmap = env['STATUSPAGE_TEST_COMPONENT_GROUP_UPTIME_ENTID']
    if ('string' === typeof idmap && idmap.startsWith('{')) {
      idmap = JSON.parse(idmap)
    }

    return { client, calls, live, idmap }
  }

  const mockFetch = async (url, init) => {
    calls.push({ url, init })
    return {
      status: 200,
      statusText: 'OK',
      headers: {},
      json: async () => (null != mockres ? mockres : { id: 'direct01' }),
    }
  }

  const client = new StatuspageSDK({
    base: 'http://localhost:8080',
    system: { fetch: mockFetch },
  })

  return { client, calls, live, idmap: {} }
}
  
